src/Eshopworld.WorkerProcess/LeaseAllocator.cs

```diff
@@ -34,83 +35,103 @@ public class LeaseAllocator : ILeaseAllocator
+                _options.Value.Priority, $"Leader election starting"));
 
-            // check lease priority is less than current worker priority and lease is not acquired
-            if (lease.Priority <= _options.Value.Priority && lease.LeasedUntil == null ||
-                lease.LeasedUntil.HasValue && lease.LeasedUntil.Value < ServerDateTime.UtcNow)
-            {
-                lease.Priority = _options.Value.Priority;
-                lease.InstanceId = instanceId;
+            var winnerInstanceId = await _leaseStore.SelectWinnerRequestAsync(_options.Value.WorkerType);
 
-                var updateResult = await _leaseStore.TryUpdateLeaseAsync(lease).ConfigureAwait(false);
+            if (!winnerInstanceId.HasValue || winnerInstanceId != instanceId)
+            {
+                var reason = !winnerInstanceId.HasValue
+                    ? "There are no LeaseRequests to evaluate"
+                    : $"Lease failed to be acquired after election. Winner instance is {winnerInstanceId.Value}.";
+                _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
+                    _options.Value.Priority, reason));
+                return null;
+            }
 
-                if (updateResult.Result)
+            if (lease == null)
+            {
+                var createTime = ServerDateTime.UtcNow;
+                var interval = _slottedInterval.Calculate(createTime, _options.Value.LeaseInterval);
+                lease = new CosmosDbLease
                 {
-                    lease = updateResult.Lease;
-
-                    // backoff to allow other worker lease instances to update the lease
-                    var delay = _allocationDelay.Calculate(_options.Value.Priority, _options.Value.LeaseInterval);
-
-                    _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
-                        _options.Value.Priority,
-                        $"Lease activation delayed [{delay}]. Allocated Lease Id: [{lease.Id}]"));
-
-                    await Task.Delay(delay).ConfigureAwait(false);
-
-                    // activate
-                    var now = ServerDateTime.UtcNow;
-                    lease.Interval = _slottedInterval.Calculate(ServerDateTime.UtcNow, _options.Value.LeaseInterval);
-                    lease.LeasedUntil = now.Add(lease.Interval.Value);
-
-                    updateResult = await _leaseStore.TryUpdateLeaseAsync(lease).ConfigureAwait(false);
+                    InstanceId = instanceId,
+                    Interval = interval,
+                    LeasedUntil = createTime.Add(interval),
+                    Priority = _options.Value.Priority,
+                    LeaseType = _options.Value.WorkerType
+                };
+                var createResult = await TryAcquireLease(lease, instanceId, _leaseStore.TryCreateLeaseAsync).ConfigureAwait(false);
+                if (createResult.Result || createResult.Lease==null)
```

Comment:
Can you clarify why this may happen?
Also, `leaseStore` is not acquiring lease any more, it simply persists whatever you give it to persist. And even when it fails to create/update the document because it was already updated/created, it should return 'ok' with the lease details.

Reply:
resolved as per discussion below